packages/@uppy/utils/src/generateFileID.js

```diff
@@ -7,15 +7,27 @@
  *
  */
 module.exports = function generateFileID (file) {
-  // filter is needed to not join empty values with `-`
-  return [
-    'uppy',
-    file.name ? encodeFilename(file.name.toLowerCase()) : '',
-    file.type,
-    file.meta && file.meta.relativePath ? encodeFilename(file.meta.relativePath.toLowerCase()) : '',
-    file.data.size,
-    file.data.lastModified
-  ].filter(val => val).join('-')
+  let id = 'uppy'
+  if (typeof file.name === 'string') {
+    id += '-' + encodeFilename(file.name.toLowerCase())
+  }
+
+  if (file.type !== undefined) {
+    id += '-' + file.type
+  }
+
+  if (file.meta && typeof file.meta.relativePath === 'string') {
+    id += '-' + encodeFilename(file.meta.relativePath.toLowerCase())
+  }
+
+  if (file.data.size !== undefined) {
+    id += '-' + file.data.size
+  }
+  if (file.data.lastModified !== undefined) {
+    id += '-' + file.data.lastModified
+  }
+
+  return id
```

Comment:
Wow, impressive it's faster than the previous solution!
Maybe let's add a comment explaining why we went for this solution? I imagine I'd look at it and try to rewrite it in a more concise way.